### PR TITLE
feat: support versioned migration

### PR DESCRIPTION
**Pull request by giautm**

### CODE REVIEW

File: docs/resources/migration.md [hunk 1]
```diff
@@ -0,0 +1,44 @@
...
+## Schema
+
+### Required
+
+- `dir` (String) the URL of the migration directory, by default it is file://migrations, e.g a directory named migrations in the current working directory.
+- `url` (String, Sensitive) The url of the database see https://atlasgo.io/cli/url
+
+### Optional
+
+- `dev_url` (String, Sensitive) The url of the dev database see https://atlasgo.io/dev-database
```
Comment (rotemtam) [inline docs/resources/migration.md:25]: Not needed for now
```suggestion
```

File: internal/atlas/atlas.go [hunk 2]
```diff
@@ -0,0 +1,214 @@
...
+	var report StatusReport
+	if err := c.runCommand(ctx, args, &report); err != nil {
+		return nil, err
+	}
+	return &report, nil
+}
+
+func (c *MigrateClient) runCommand(ctx context.Context, args []string, report interface{}) error {
+	cmd := exec.CommandContext(ctx, c.path, args...)
+	cmd.Env = append(cmd.Env, "ATLAS_NO_UPDATE_NOTIFIER=1")
```
Comment (rotemtam) [inline internal/atlas/atlas.go:115]: 👍

File: internal/atlas/atlas_test.go [hunk 3]
```diff
@@ -0,0 +1,135 @@
...
+	"ariga.io/ariga/terraform-provider-atlas/internal/atlas"
+	"ariga.io/atlas/sql/sqlclient"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	mysqlURL = "mysql://root:pass@localhost:3306"
+)
+
+func Test_migrateApply(t *testing.T) {
```
Comment (rotemtam) [inline internal/atlas/atlas_test.go:21]:
```suggestion
func Test_MigrateApply(t *testing.T) {
```

File: internal/atlas/atlas_test.go [hunk 4]
```diff
@@ -0,0 +1,135 @@
...
+	_ "ariga.io/atlas/sql/mysql"
+	_ "github.com/go-sql-driver/mysql"
+
+	"ariga.io/ariga/terraform-provider-atlas/internal/atlas"
+	"ariga.io/atlas/sql/sqlclient"
+	"github.com/stretchr/testify/require"
+)
+
+const (
+	mysqlURL = "mysql://root:pass@localhost:3306"
```
Comment (rotemtam) [inline internal/atlas/atlas_test.go:18]: i think to simplify the test you can work against sqlite here?

File: internal/atlas/atlas_test.go [hunk 5]
```diff
@@ -0,0 +1,135 @@
...
+			}
+			if !reflect.DeepEqual(got.Next, tt.wantNext) {
+				t.Errorf("migrateStatus() = %v, want %v", got.Next, tt.wantNext)
+			}
+		})
+	}
+}
+
+func tempSchemas(t *testing.T, schemas ...string) {
+	c, err := sqlclient.Open(context.Background(), mysqlURL)
```
Comment (rotemtam) [inline internal/atlas/atlas_test.go:109]: if you use sqlite, you don't need the setup/cleanup

**Review by rotemtam — COMMENTED** [review]
@giautm tremendous work, i started reviewing but tbh this is too big to review deeply. Can you please break it down to a few chunks? Thanks.

File: internal/atlas/atlas.go [hunk 6]
```diff
@@ -0,0 +1,214 @@
...
+				detail:  string(output),
+			}
+		}
+		if exitErr.ExitCode() != 1 || !json.Valid(output) {
+			return &cliError{
+				summary: "Atlas CLI",
+				detail:  string(output),
+			}
+		}
+		// When the exit code is 1, it means that the command
```
Comment (masseelch) [inline internal/atlas/atlas.go:134]: No, when it is 0. 1 (technically >0) means there is an error.

Comment (giautm) [reply internal/atlas/atlas.go:134]: @masseelch: I need to read the output of `migrate lint` which always return non-zero for case of lint error.

File: internal/atlas/atlas.go [hunk 7]
```diff
@@ -0,0 +1,201 @@
...
+	"path"
+	"runtime"
+	"strings"
+
+	"github.com/hashicorp/terraform-plugin-framework/diag"
+)
+
+type (
+	// MigrateClient is a client for the Atlas CLI.
+	MigrateClient struct {
```
Comment (rotemtam) [inline internal/atlas/atlas.go:18]: How about just `Client`? `atlas.Client` will look nice in the codebase.

File: internal/atlas/atlas.go [hunk 8]
```diff
@@ -0,0 +1,201 @@
...
+	if err != nil {
+		return nil, err
+	}
+	return NewClientWithPath(path)
+}
+
+// NewClientWithPath returns a new Atlas client with the given atlas-cli path.
+func NewClientWithPath(path string) (*MigrateClient, error) {
+	return &MigrateClient{path: path}, nil
+}
```
Comment (rotemtam) [inline internal/atlas/atlas.go:52]: don't need the `error` return value here right?

File: internal/atlas/atlas.go (hunk 6 above)
Comment (rotemtam) [reply internal/atlas/atlas.go:120]: Are you sure? IIRC non-zero codes are returned only if we have diagnostics that should cause the build to crash. e.g, data-dependent changes still return diagnostics but return status 0.

File: internal/atlas/atlas_test.go [hunk 9]
```diff
@@ -0,0 +1,135 @@
...
+	c, err := atlas.NewClient("atlas")
+	r.NoError(err)
+	for _, tt := range tests {
+		t.Run(tt.name, func(t *testing.T) {
+			got, err := c.Status(tt.args.ctx, tt.args.data)
+			if (err != nil) != tt.wantErr {
+				t.Errorf("migrateStatus() error = %v, wantErr %v", err, tt.wantErr)
+				return
+			}
+			if !reflect.DeepEqual(got.Current, tt.wantCurrent) {
```
Comment (rotemtam) [inline internal/atlas/atlas_test.go:98]: i think. you can use `require.EqualValues` ?

File: internal/provider/atlas_migration_resource.go [hunk 10]
```diff
@@ -0,0 +1,320 @@
...
+	if data.Version.IsNull() {
+		amount = len(statusReport.Pending)
+	} else {
+		for idx, v := range statusReport.Pending {
+			if v.Version == data.Version.Value {
+				amount = idx + 1
+				break
+			}
+		}
+		if amount == 0 {
```
Comment (rotemtam) [inline internal/provider/atlas_migration_resource.go:250]: we shouldn't fail if the latest migration is already applied be a previous `terraform apply`

**Review by rotemtam — COMMENTED** [review]
Amazing job. see comments

File: internal/atlas/atlas.go (hunk 6 above)
Comment (a8m) [reply internal/atlas/atlas.go:120]: It depends on the configuration. See the second block here: https://atlasgo.io/lint/analyzers#destructive-changes

File: internal/atlas/atlas_test.go (hunk 5 above)
Comment (giautm) [reply internal/atlas/atlas_test.go:102]: I'm not sure. Because the existing code using MySQL. I will migrate to sqllite in another PR if needs

**Review by rotemtam — APPROVED** [review]
Wonderful job, @giautm. This is very exciting

**Pull request merged**